Work Generation page

zhyatt opened the pull request:
Added new Integration Guides > Work Generation page with various new details and some existing things pulled over from Integration Guides > The Basics page.

- [x] Create ticket for a future PR to ensure base difficulty values are updated https://github.com/nanocurrency/nano-docs/issues/253

guilhermelawless commented on docs/integration-guides/work-generation.md:
```diff
@@ -0,0 +1,209 @@
+
+**Development node wallet behavior**
+
+The developer wallet included with the node is configured to pre-cache work as the base threshold and monitor any blocks it publishes to the network for confirmation. If they are not confirmed within 5 seconds, the difficulty on that block will be compared against the active difficulty seen on the network. If the block has a lower work value than the network then new work generation is requested  ...
```
We can also mention there is a config for the maximum multiplier (again, we'll have to decide which difficulty that will be based off of)

guilhermelawless replied:
We do use it outside of local work generation, any requests to peers should never be above that maximum multiplier (more concretely, the equivalent difficulty). This is to prevent issues with the node wallet trying to generate work at higher difficulty but this never being completed.

zhyatt replied:
Ok, I think I updated this right and included as a notable config callout linked in the update.

guilhermelawless replied:
Also of note for external integrations (not only the dev wallet) since the RPC will reject requests for difficulty levels above this maximum.

zhyatt replied:
Moved to a tip callout box below external integrations

guilhermelawless commented on docs/integration-guides/work-generation.md:
```diff
@@ -0,0 +1,209 @@
+!!! warning "Upcoming threshold changes and variations by block type"
+	Plans are underway to change the thresholds based on the type of block with the release of V21 and subsequent distribution of v2 epoch blocks to enable the feature. See the [Development Update: V21 PoW Difficulty Increases article](https://medium.com/nanocurrency/development-update-v21-pow-difficulty-increases-362b5d052c8e) for full details.
+
+### Pre-caching
```
Should we note this is handled automatically by the dev wallet?

guilhermelawless commented on docs/snippets/alternative-work-generation-setup-preferred.md:
```diff
@@ -0,0 +1,2 @@
+!!! warning "Alternative work generation setup preferred"
+	This option is not recommended for work generation for nodes due to the potential for performance impacts to the node participating on the network (voting, staying in sync, etc.). Please see the [Work Generation guide](/integration-guides/work-generation) for best practices.
```
> This option is not recommended

I think we should make it more clear that the recommendation is to set to "0" in these cases, which is not the default.

guilhermelawless commented on docs/integration-guides/work-generation.md:
```diff
@@ -0,0 +1,209 @@
+
+To provide a more robust and redundant work generation setup, multiple [work peers](/running-a-node/configuration/#nodework_peers) can be used. Any node configured with multiple peers will make requests serially from the list of work peers until a successful response is received.
+
+!!! tip "Disable local CPU work generation"
```
It was related to the work_peers config: https://docs.nano.org/running-a-node/configuration/#nodework_peers . `"::ffff:127.0.0.1:7076"` is the local node

zhyatt replied:
Updated

guilhermelawless commented on docs/integration-guides/work-generation.md:
```diff
@@ -0,0 +1,209 @@
+
+Services with heavy RPC calls and work generation can benefit from ensuring dedicated resources exist for each process separately. To maximize performance a separate machine running the [Nano Work Server](https://github.com/nanocurrency/nano-work-server) with a GPU attached is recommended:
+
+1. Setup a machine separate from the node with GPU attached
```
Yes, and being aware of anything that could delay the connection (things like Cloudflare perhaps, although that is not supported currently)

zhyatt replied:
Added callout in the "Choosing a machine" section further up

guilhermelawless commented on docs/integration-guides/work-generation.md:
```diff
@@ -0,0 +1,209 @@
+	* CPU: `nano_node --debug_profile_generate`
+	* GPU: `nano_node --debug_opencl --platform=0 --device=0` (updating `platform` and `device` as necessary)
+
+This will trigger continual work generation, so let it run until a sufficient sample size of times are generated (at least 100 instances). Compute the average of these times which is the number of microseconds to generate work.
```
Agreed

zhyatt commented:
@guilhermelawless  Also, on a more broad note, do you think this organization/approach overall is the best way to set things up? I know we are already deep into revising but open to shifting things around or framing different if needed.

guilhermelawless commented:
I personally like the current approach, and we can iterate in the future based on feedback. It's possible that dividing at the top-most level by internal dev wallet and external integration would be better, but would lead to a lot of redundancy.

zhyatt commented:
@guilhermelawless I moved some of the previous items to be considered for this PR out to their own tickets for later consideration and adjusted some final things. Any last changes before this is pushed out?

guilhermelawless commented on docs/integration-guides/work-generation.md:
```diff
@@ -137,14 +137,15 @@ The mainnet's base difficulty threshold is currently `0xffffffc000000000` for al
 For services aiming to ensure the highest priority on their transactions, the confirmation of published blocks should be monitored by their integration and work levels compared against active difficulty in a similar fashion to the development wallet mentioned above. If work is left at base difficulty there could be delays in the transactions being processed during heavy network usage times.
 
+!!! tip "Configure max work generate multiplier"
+    Due to the possibility of network work levels increasing beyond the capabilities of certain work generation setups, the config option [`node.max_work_generate_multiplier`](/running-a-node/configuration/#nodework_threads) can be used to limit how high a work value will be requested at. All setups, whether using the developer wallet or an external integration, should implement an appropriate li ...
```
```suggestion
    Due to the possibility of network work levels increasing beyond the capabilities of certain work generation setups, the config option [`node.max_work_generate_multiplier`](/running-a-node/configuration/#nodemax_work_generate_multiplier) can be used to limit how high a work value will be requested at. All setups, whether using the developer wallet or an external integration, should implement an appropriate limit which defaults to 64x in V20.
```

guilhermelawless replied:
Other than this looks good to me 👍

zhyatt commented:
Thanks for your help on this @guilhermelawless !
